File: training/src/compiler/training/base/layers/basic/LabelSmoothing.cpp
// Copyright (C) 2022. Huawei Technologies Co., Ltd. All rights reserved.

// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#include "LabelSmoothing.h"

#include <algorithm>
#include <limits>

#include <training/base/common/MemoryManager.h>

#include "impl/LabelSmoothingCPU.h"

namespace
{

const size_t NoPadding = std::numeric_limits<std::size_t>::max();

} // anonymous namespace

namespace raul
{

LabelSmoothing::LabelSmoothing(const Name& name, const LabelSmoothingParams& params, NetworkParameters& networkParameters)
    : BasicLayer(name, "LabelSmoothing", params, networkParameters)
    , mSmoothing(TODTYPE(params.smoothing))
    , mPaddingIdx(params.paddingClass >= 0 ? params.paddingClass : NoPadding)
{
    auto prefix = mTypeName + "[" + mName + "::ctor]: ";

    if (mInputs.size() != 1)
    {
        THROW(mTypeName, mName, "wrong number of input names");
    }
    if (mOutputs.size() != 1)
    {
        THROW(mTypeName, mName, "wrong number of output names");
    }

    DECLARE_IMPL(LabelSmoothing, LabelSmoothingCPU<MemoryManager>, LabelSmoothingCPU<MemoryManagerFP16>)

    mInputName = mInputs[0];
    mOutputName = mOutputs[0];

    if (mNetworkParams.mWorkflow.getWidth(mInputName) < 2)
    {
        THROW(mTypeName, mName, "input width must be at least 2");
    }
    if (mNetworkParams.mWorkflow.getWidth(mInputName) == 2 && mPaddingIdx != NoPadding)
    {
        THROW(mTypeName, mName, "when using padding, input width must be at least 3");
    }

    mNetworkParams.mWorkflow.copyDeclaration(mName, mInputName, raul::Workflow::Usage::Forward, raul::Workflow::Mode::Read);

    mNetworkParams.mWorkflow.copyDeclaration(mName, mInputName, mOutputName, DEC_FORW_WRIT);
}

} // namespace raul

File: training/src/tests/tests/python/labelSmoothing.py
# Copyright (C) 2022. Huawei Technologies Co., Ltd. All rights reserved.

# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


import torch
import torch.nn as nn
import numpy as np
import torchvision
from torch.autograd import Variable


class LabelSmoothing(nn.Module):
    "Implement label smoothing."

    def __init__(self, size, padding_idx, smoothing=0.0):
        super(LabelSmoothing, self).__init__()
        self.criterion = nn.KLDivLoss(reduction="batchmean")
        self.padding_idx = padding_idx
        self.confidence = 1.0 - smoothing
        self.smoothing = smoothing
        self.size = size
        self.true_dist = None

    def forward(self, x, target):
        assert x.size(1) == self.size
        true_dist = x.data.clone()
        true_dist.fill_(self.smoothing / (self.size - 2))
        true_dist.scatter_(1, target.data.unsqueeze(1), self.confidence)
        true_dist[:, self.padding_idx] = 0
        mask = torch.nonzero(target.data == self.padding_idx)
        if mask.dim() > 0:
            true_dist.index_fill_(0, mask.squeeze(), 0.0)
        self.true_dist = true_dist
        print("smoothed", true_dist)
        return self.criterion(x, Variable(true_dist, requires_grad=False))


m = nn.LogSoftmax(dim=1)
loss = LabelSmoothing(4, 0, 0.1)
inp = torch.tensor(
    [
        [1.0, 2.0, 5.0, -1.0],
        [1.0, 2.0, 6.0, -1.0],
        [-2.0, 5.0, -1.0, 3.0],
    ],
    requires_grad=True,
)
target = torch.tensor([1, 2, 0])
t1 = m(inp)
t1.retain_grad()
output = loss(t1, target)
print("loss", output)
output.backward()
print(t1)
print(t1.grad)

print(inp.grad)
